apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,167 @@
+            }
+
+            final HandlerType handlerType = (HandlerType) handlerTypeField.get(ctx);
+            final String handlerClassName = handler.getClass().getName();
+
+            // transaction name like the HTTP path
+            if (handlerType.isHttpMethod()) {
+                transaction.setFrameworkName(FRAMEWORK_NAME);
+                transaction.setFrameworkVersion(VersionUtils.getVersion(Handler.class, "io.javalin", "javalin"));
+                // no need for anonymous handler class names in the transaction
+                if (handlerClassName.startsWith("co.elastic.apm.agent.javalin.JavalinHandlerLambdaInstrumentation$WrappingHandler")
+                    || handlerClassName.startsWith("io.javalin.http.JavalinServlet")) {
+                    setName(transaction, handlerType.name() + " " + ctx.endpointHandlerPath());
+                } else {
+                    setName(transaction, handlerType.name() + " " + ctx.endpointHandlerPath() + " " + handlerClassName);
+                }
+            }
+
+            // do not create an own span for JavalinServlet.addHandler, as this not added by the users code and leads to confusion
+            if (handlerClassName.startsWith("io.javalin.http.JavalinServlet")) {
+                return null;
+            }
+
+            // create own span for all handlers including after/before
+            final Span span = startSpan(handlerType.name(), ctx.matchedPath(), ctx.method(), ctx.url());
+            return span;
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void onAfterExecute(@Advice.Enter @Nullable Object spanObj,
+                                          @Advice.Argument(0) Object ctxObject,
+                                          @Advice.Thrown @Nullable Throwable t) {
+            if (spanObj instanceof Span) {
+                Context ctx = ((Context) ctxObject);
+                final Span span = (Span) spanObj;
+                span.deactivate();
+
+                final CompletableFuture<?> responseFuture = ctx.resultFuture();
+                if (responseFuture == null) {
+                    // sync request
+                    span.captureException(t).end();
+                } else {
+                    // future was set in the handler, so we need to end the span only on future completion
+                    responseFuture.whenComplete((o, futureThrowable) -> span.captureException(futureThrowable).end());
+                }
+            }
+        }
+
+        private static Span startSpan(String handlerTypeName, String matchedPath, String method, String url) {
+            final AbstractSpan<?> parent = tracer.getActive();
+            if (parent == null) {
+                return null;
+            }
+
+            Span span = parent.createSpan().activate();
+            // TODO figure out correct type and subtype or leave out?
+            span.withType("web")
```

Comment:
We are soon to have a [spec](https://github.com/elastic/apm/pull/443/files#diff-b09b7eae35b49561ec883168d53a9b112c63632aee03d2977ccf81f300bc614e) for allowed types and subtypes and since this is not part of it, the test will fail.
If you can't find something relevant in the spec, you may propose adding this.

Reply:
@SylvainJuge WDYT would be best?

Reply:
That's a very good question !

The other example we have for such internal span types is the one for `servlet` / `request-dispatcher`.
If we follow the same pattern we could pick `javalin` / `javalin-handler`.

Would there exist other span subtypes ?

Reply:
not in this case currently, will go with `javalin/javalin-handler` then